### DISCUSSION TITLE
Running the server

### DISCUSSION THREAD

**Discussion by samysosaw** [Hucaru/Valhalla#50] [Q&A]
Hi, I would like to run the server to learn/practice Go with a game I love and was wondering if we could get more details as to how to run this ?

Here are the steps I followed:

- Installed Docker and Docker-Compose
- Downloaded the Data.nx and added it at the root of the repo
- Downloaded the repo and ran docker compose up
- Everything started without any issues
- Downloaded the v28_noaes.exe client
- Downloaded the v28 game files and put the client inside
- Launched the client and everything worked

Here is my problem. When I try to login, I can't seem to login. I tried connecting to the valhalla-master-db-1 container with root and inserting a row in the accounts table using a query like this : insert into accounts (accountID, username, password, isLogedIn, adminLevel, isBanned, gender, dob) values (1, "user123", "test123", 0, 0, 0, 0, 1990);

The query executed fine, however I still can't log in using the provided info (user123/test123). I get the "Invalid login id" or the "wrong password" errors.  The client seems to be connecting to the server because I can see the following output : 2022/05/26 14:19:25 New client from 172.X.X.X:56744

Is there anything else to get this working that I missed ? 

Thank you for your help !

**Comment by ErwinsExpertise** [comment]
Are you hashing the password? The source expects an MD5 hashed password, not plaintext.

**Reply by samysosaw** [reply]
Do you mean something like this ?

insert into accounts (accountID, username, password, isLogedIn, adminLevel, isBanned, gender, dob) values (6, "testuser", "482c811da5d5b4bc6d497ffa98491e38", 0, 0, 0, 0, 1990);

I tried that too (MD5 hash = password123) and im still getting the wrong password error

Also tried insert into accounts (accountID, username, password, isLogedIn, adminLevel, isBanned, gender, dob) values (8, "testuser2", MD5('password123'), 0, 0, 0, 0, 1990);

I also tried creating the user through the Adminer interface, same result.

Could it be related to this bit of code ? 

	hasher := sha512.New()
	hasher.Write([]byte(password))
	hashedPassword := hex.EncodeToString(hasher.Sum(nil))

**Reply by samysosaw** [reply]
I actually got it working by removing this bit of code and replacing this line with this line. I know this removes the security benefits, but since I won't be using it to play online or anything it will be fine.

else if hashedPassword != databasePassword {
		result = 0x04
	} 

else if password != databasePassword {
		result = 0x04
	}

**Reply by Hucaru** [reply]
It's as Erwin says. I suspect you are not hashing the password correctly.

**Reply by ErwinsExpertise** [reply]
@samoomoo I apologize, I was incorrect in that reply. It actually expects SHA512 hash not MD5. An easy way to get the exact hash you would need is with the following:
https://go.dev/play/p/61jm6PJwxL3

Also @Hucaru I'm going to try to get back into working on the source again. Hopefully I can get some more of it working

**Reply by Hucaru** [reply]
@ErwinsExpertise go for it! I currently don't have the bandwidth for doing any development but I can review pull requests when they come in.

**Reply by samysosaw** [reply]
I used a SHA512 hash and it worked, thanks a lot !

One more issue, but this time it's with the localhost. I don't know much about editing clients so forgive me for the noob question. I tried to search the address 0066519c in the v28_noaes.exe file and it seems to stop at 0038A1F8 for some reason (using HxD). Is there anything im not doing right ? I would like to have the game in windowed mode. Thanks !

**Reply by ErwinsExpertise** [reply]
I'm not too familiar with client edits so can't help you there, but it seems like you are on this. I could use some help working on the source if you are able to.

**Reply by Hucaru** [reply]
The client offsets where discovered within a debugger which means they will be offset from the start of the executable. I suggest making the edits in an application like x32dbg

**Reply by samysosaw** [reply]
Yeah I would be interested ! Im still a beginner in Go and don't have much free time to work on it, but this looks like a good project to learn. The only issue with client edits is that most of the documentation online is for more recent versions (like v83 and up). 

Maybe @Hucaru could have more info to run the game in windowed mode since theres info about that in the readme.

**Reply by Hucaru** [reply]
load the localhost in this https://x64dbg.com/ and use the offsets in the readme. The offsets won't work if you use a hex editor as I haven't normalised them. If they still don't work then use the maplestory client hook I made https://github.com/Hucaru/maplestory-client-hook

The offset for window mode is 26459C and should start with the a.o.b. C745 E4 10000000 but I suggest using a debugger to edit the application as it will take care of any padding for you. Using the difference between the value I gave here and the readme you can calculate the other offsets.

**Reply by samysosaw** [reply]
Thanks for the info ! I was able to make it windowed and remove the popup after closing. EDIT : Had to remove it from Google since it's flagged as containing a virus. If anyone needs it feel free to message me and i'll find a way to share it.
